lib/pure/strinterp.nim

```diff
@@ -0,0 +1,320 @@
+#
+#
+#            Nim's Runtime Library
+#        (c) Copyright 2012 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## This module provides the string interpolation macro ``fmt``.
+
+import parseutils, sequtils, macros, strutils
+
+# -----------------------------------------------------------------------------
+# nimboost's formatters
+# -----------------------------------------------------------------------------
+
+proc mkDigit(v: int, lowerCase: bool): string {.inline.} =
+  doAssert(v < 26)
+  if v < 10:
+    result = $chr(ord('0') + v)
+  else:
+    result = $chr(ord(if lowerCase: 'a' else: 'A') + v - 10)
+
+proc intToStr(n: SomeNumber, radix = 10, len = 0, fill = ' ', lowerCase = false): string =
+  ## Converts ``n`` to string. If ``n`` is `SomeReal`, it casts to `int64`.
+  ## Conversion is done using ``radix``. If result's length is lesser then
+  ## ``len``, it aligns result to the right with ``fill`` char.
+  ## If ``len`` is negative, the result is aligned to the left.
+  ## If `lowerCase` is true, formatted string will be in the lower case.
+  when n is SomeUnsignedInt:
+    var v = n.uint64
+    let s = false
+  else:
+    var v = n.int64
+    let s = v.int64 < 0
+    if s:
+      v = v * -1
+
+  if v == 0:
+    result = "0"
+  else:
+    result = ""
+    while v > (type(v))0:
+      let d = v mod (type(v))radix
+      v = v div (type(v))radix
+      result.add(mkDigit(d.int, lowerCase))
+    for idx in 0..<(result.len div 2):
+      swap result[idx], result[result.len - idx - 1]
+  var length = abs(len)
+  if length == 0 or (s and (result.len >= length - 1)) or (not s and (result.len >= length)):
+    if s:
+      result = "-" & result
+  elif len < 0:
+    if s:
+      result = "-" & result
+    for i in result.len..<length:
+      result.add(fill)
+  else:
+    if fill != '0':
+      # The sign must be near the number
+      if s:
+        result = "-" & result
+    var toFill = length - result.len
+    var prefix = newString(toFill)
+    for idx in 0..<toFill:
+      prefix[idx] = fill
+    if fill == '0' and s:
+      prefix[0] = '-'
+    result = prefix & result
+
+proc alignStr(s: string, len: int, fill = ' ', trunc = false): string =
```

Comment:
To me this function seems to do too much, but maybe it makes the implementation easier. Might be worth considering splitting this up into a ``truncate`` and ``indent`` function.

Reply:
Also, we do have an ``indent`` in the strutils module already, could it be used? 

https://nim-lang.org/docs/strutils.html#indent,string,Natural,string

Reply:
Yes, it would be possible, but from the perspective of the macro it is nicer to have just one function to call which does all expected behavior at once, instead of having to build a more complex AST with chaining function calls. That's why I though it helps to rename these functions.